Document TestingPlatformBuilderHook auto-registration for MTP extension authors (#54122)
dotnet/docs: docs/core/testing/microsoft-testing-platform-architecture-extensions.md

* `ExitCode`: The exit code of the process. This value is only available within the `OnTestHostProcessExitedAsync` method. Attempting to access it within the `OnTestHostProcessStartedAsync` method will result in an exception.
* `HasExitedGracefully`: A boolean value indicating whether the test host has crashed. If true, it signifies that the test host did not exit gracefully.

## Auto-register your extension with `TestingPlatformBuilderHook`

> Every extension subsection above shows a *manual* registration call (for example, `builder.TestHost.AddDataConsumer(...)`). Asking consumers to edit their `Main` method is a poor onboarding experience. The [`Microsoft.Testing.Platform.MSBuild`](https://www.nuget.org/packages/Microsoft.Testing.Platform.MSBuild) package solves this by generating a `SelfRegisteredExtensions.AddSelfRegisteredExtensions(builder, args)` method that runs from the auto-generated entry point. To plug your extension into that generated method, ship two artifacts in your NuGet package:

Reviewer: ```suggestion
Every preceding extension section shows a *manual* registration call (for example, `builder.TestHost.AddDataConsumer(...)`). Asking consumers to edit their `Main` method is a poor onboarding experience. The [`Microsoft.Testing.Platform.MSBuild`](https://www.nuget.org/packages/Microsoft.Testing.Platform.MSBuild) package solves this by generating a `SelfRegisteredExtensions.AddSelfRegisteredExtensions(builder, args)` method that runs from the autogenerated entry point. To plug your extension into that generated method, ship two artifacts in your NuGet package:
```

Reviewer: We're supposed to avoid directional terms.

1. A public static `TestingPlatformBuilderHook` class with an `AddExtensions` method that registers your extension.

> 2. An MSBuild props file that declares a `<TestingPlatformBuilderHook>` item pointing at that class.

Reviewer: ```suggestion
- A public static `TestingPlatformBuilderHook` class with an `AddExtensions` method that registers your extension.
- An MSBuild props file that declares a `<TestingPlatformBuilderHook>` item pointing at that class.
```

Reviewer: Ordered lists are only supposed to be used for ordered steps.

> When a consumer installs your package, the MSBuild integration picks up the item, generates the call into your hook, and your extension is registered with no code changes on the consumer side.

Reviewer: ```suggestion
When a consumer installs your package, the MSBuild integration picks up the item and generates the call into your hook, and your extension is registered with no code changes on the consumer side.
```

> [!NOTE]

> > Auto-registration only works when the consumer has `Microsoft.Testing.Platform.MSBuild` in their project (it's included transitively by MSTest, NUnit, and xUnit runners) and hasn't opted out by setting `<GenerateTestingPlatformEntryPoint>false</GenerateTestingPlatformEntryPoint>`. Consumers who disable the auto-generated entry point still need to call your manual registration API from their `Main` method.

Reviewer: ```suggestion
> Auto-registration only works when the consumer has `Microsoft.Testing.Platform.MSBuild` in their project (it's included transitively by MSTest, NUnit, and xUnit runners) and hasn't opted out by setting `<GenerateTestingPlatformEntryPoint>false</GenerateTestingPlatformEntryPoint>`. Consumers who disable the autogenerated entry point still need to call your manual registration API from their `Main` method.
```

### Create the hook class

Add a `public static class TestingPlatformBuilderHook` in your extension assembly with an `AddExtensions(ITestApplicationBuilder, string[])` method that performs the same registration users would otherwise call manually:

```csharp
using Microsoft.Testing.Platform.Builder;

namespace Contoso.MyExtension;

public static class TestingPlatformBuilderHook
{
    public static void AddExtensions(ITestApplicationBuilder testApplicationBuilder, string[] arguments)
        => testApplicationBuilder.AddMyExtension();
}
```

The class name doesn't have to be `TestingPlatformBuilderHook` — the MSBuild item points at it by full type name — but using that name keeps your code consistent with the in-box extensions like `Microsoft.Testing.Extensions.Retry` and `Microsoft.Testing.Extensions.HotReload`.

> The method signature must match exactly:

Reviewer: ```suggestion
The method must:
```

* It must be `public static`.
* The first parameter must be `Microsoft.Testing.Platform.Builder.ITestApplicationBuilder`.
* The second parameter must be `string[]` (the command-line arguments passed to the test host). You can ignore it if your extension doesn't need it.

> * It must return `void`.

Reviewer: ```suggestion
* Be `public static`.
* Have a first parameter of type `Microsoft.Testing.Platform.Builder.ITestApplicationBuilder`.
* Have a second parameter of type `string[]` (the command-line arguments passed to the test host). You can ignore it if your extension doesn't need it.
* Return `void`.
```

### Declare the MSBuild item

Ship a props file under `buildMultiTargeting/<PackageId>.props` inside your NuGet package. Declare a `<TestingPlatformBuilderHook>` item that points the MSBuild task at your hook class:

```xml
<Project>
  <ItemGroup>
    <TestingPlatformBuilderHook Include="xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx">
      <DisplayName>Contoso.MyExtension</DisplayName>
      <TypeFullName>Contoso.MyExtension.TestingPlatformBuilderHook</TypeFullName>
    </TestingPlatformBuilderHook>
  </ItemGroup>
</Project>
```

> The metadata is:

Reviewer: ```suggestion
The metadata is as follows:
```

* `Include`: A GUID that uniquely identifies your hook. See [The `Include` GUID is a random identifier](#the-include-guid-is-a-random-identifier).
* `DisplayName`: The friendly name shown in MSBuild diagnostic messages when the entry point is generated. Use your package or extension name.

> * `TypeFullName`: The fully qualified name of the `TestingPlatformBuilderHook` class you created above. The MSBuild task uses this to emit `global::Contoso.MyExtension.TestingPlatformBuilderHook.AddExtensions(builder, args);` into the generated entry point.

Reviewer: ```suggestion
* `TypeFullName`: The fully qualified name of the `TestingPlatformBuilderHook` class you created previously. The MSBuild task uses this to emit `global::Contoso.MyExtension.TestingPlatformBuilderHook.AddExtensions(builder, args);` into the generated entry point.
```
